### PR TITLE
Add support for HS4XD SoC

### PR DESCRIPTION
**Pull request by kokas-a**
Add set of initial files for HS4XD SoC support. Based on HSDK sources.

### CODE REVIEW

File: boards/arc/hs4xd/hs4xd_defconfig [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_SOC_ARC_HS4XD=y
+CONFIG_BOARD_HS4XD=y
+CONFIG_SYS_CLOCK_TICKS_PER_SEC=100
+CONFIG_XIP=y
```
Comment (evgeniy-paltsev) [inline boards/arc/hs4xd/hs4xd_defconfig]: Any reason to have XIP enabled?

We don't have any XIP flash memory and we don't have XIP enabled for HSDK v1.

File: soc/arc/snps_arc_hs4xd/Kconfig.defconfig [hunk 2]
```diff
@@ -0,0 +1,50 @@
+# Copyright (c) 2019 Synopsys, Inc. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+if SOC_ARC_HS4XD
+
+config SOC
+	default "snps_arc_hs4xd"
```
Comment (evgeniy-paltsev) [inline soc/arc/snps_arc_hs4xd/Kconfig.defconfig]: snps_arc_hs4xd -> snps_arc_hsdk4xd

File: boards/arc/hs4xd/Kconfig.defconfig [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
+
+if BOARD_HS4XD
+
+config BOARD
+	default "hs4xd"
```
Comment (evgeniy-paltsev) [inline boards/arc/hs4xd/Kconfig.defconfig]: hs4xd -> hsdk4xd

File: boards/arc/hs4xd/Kconfig.board [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# DesignWare ARC HS4XD Development Kit board configuration
```
Comment (evgeniy-paltsev) [inline boards/arc/hs4xd/Kconfig.board]: HS4XD -> HSDK4xD

File: boards/arc/hs4xd/Kconfig.board [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+
+# Copyright (c) 2022 Synopsys, Inc. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_HS4XD
```
Comment (evgeniy-paltsev) [inline boards/arc/hs4xd/Kconfig.board]: BOARD_HS4XD -> BOARD_HSDK4XD

File: soc/arc/snps_arc_hs4xd/Kconfig.defconfig [hunk 6]
```diff
@@ -0,0 +1,50 @@
+# Copyright (c) 2019 Synopsys, Inc. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+if SOC_ARC_HS4XD
```
Comment (evgeniy-paltsev) [inline soc/arc/snps_arc_hs4xd/Kconfig.defconfig]: SOC_ARC_HS4XD -> SOC_ARC_HSDK4XD

File: soc/arc/snps_arc_hs4xd/Kconfig.defconfig [hunk 7]
```diff
@@ -0,0 +1,50 @@
+# Copyright (c) 2019 Synopsys, Inc. All rights reserved.
```
Comment (evgeniy-paltsev) [inline soc/arc/snps_arc_hs4xd/Kconfig.defconfig]: 2019 -> current year

File: soc/arc/snps_arc_hs4xd/linker.ld [hunk 8]
```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright (c) 2019 Synopsys, Inc. All rights reserved.
```
Comment (evgeniy-paltsev) [inline soc/arc/snps_arc_hs4xd/linker.ld]: ditto

File: soc/arc/snps_arc_hs4xd/soc.c [hunk 9]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright (c) 2019 Synopsys, Inc. All rights reserved.
```
Comment (evgeniy-paltsev) [inline soc/arc/snps_arc_hs4xd/soc.c]: ditto

File: soc/arc/snps_arc_hs4xd/soc.c [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+#include <zephyr/arch/arc/v2/arc_connect.h>
+
+static int arc_hsdk_init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	uint32_t core;
+	uint32_t i;
+
+	/* allocate all IDU interrupts to master core */
```
Comment (evgeniy-paltsev) [inline soc/arc/snps_arc_hs4xd/soc.c]: master -> primary

File: soc/arc/snps_arc_hs4xd/soc.c [hunk 11]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright (c) 2019 Synopsys, Inc. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * This module provides routines to initialize and support soc-level hardware
+ * for the HS Development Kit
```
Comment (evgeniy-paltsev) [inline soc/arc/snps_arc_hs4xd/soc.c]: Copypaste, please update board name

File: soc/arc/snps_arc_hs4xd/Kconfig.soc [hunk 12]
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2019 Synopsys, Inc. All rights reserved.
```
Comment (evgeniy-paltsev) [inline soc/arc/snps_arc_hs4xd/Kconfig.soc]: ditto

File: boards/arc/hs4xd/hs4xd.dts [hunk 13]
```diff
@@ -0,0 +1,14 @@
+/*
+ * Copyright (c) 2019, Synopsys, Inc. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/dts-v1/;
+
+#include "hs4xd.dtsi"
```
Comment (evgeniy-paltsev) [inline boards/arc/hs4xd/hs4xd.dts]: hs4xd.dtsi -> hsdk4xd.dtsi

**Review by evgeniy-paltsev — CHANGES_REQUESTED** [review]
Some minor issues.

I'll check how this implementation maps to current DSP support for EM development - probably we can combine dsp implementations.

**Comment by ruuddw** [comment]
Few questions:

- Do we need a new CPU type for this? HS4x is already covered with the CPU_HS3X option. "If y, the SoC uses an ARC HS3x or HS4x CPU". There is no HS3XD either, it is simply configured by enabling 'DSP' for the HS3X
- Do we need a new board? Is there a physical board for this, or is it another config of nsim/qemu?

**Comment by evgeniy-paltsev** [comment]
Hi @ruuddw,

> Do we need a new CPU type for this? HS4x is already covered with the CPU_HS3X option. "If y, the SoC uses an ARC HS3x or HS4x CPU". There is no HS3XD either, it is simply configured by enabling 'DSP' for the HS3X

I'm not fully sure about this. The one meaningful part of having CPU_HS4X is to provide generic hs4x mcpu for all GNU toolchain targets (it's currently missing in this PR). However, as we have now possibility to tweak mcpu on SoC level - that might be a bit excessive :)

And yep, if we introduce new cpu it should be CPU_HS4X + DSP option enabled and not CPU_HS4XD as we have now.

> Do we need a new board? Is there a physical board for this, or is it another config of nsim/qemu?

Yep, it's physical board.
See "ARC HS4x/HS4xD Development Kit" section on https://www.synopsys.com/dw/ipdir.php?ds=arc-hs-development-kit page

**Comment by ruuddw** [comment]
I considered the HSDK a single board, but you're right there are two revisions. Doesn't the existing HSDK cover both revisions?
Still (I think I mentioned this at that time), using CPU_HS3x for either 3X or 4X is confusing, so I don't mind adding 4X explicitly. But let's be consistent on the "D" as option and not a main CPU.

File: soc/arc/snps_arc_hs4xd/Kconfig.defconfig (hunk 2 above)
Comment (kokas-a) [reply soc/arc/snps_arc_hs4xd/Kconfig.defconfig]: SoC based on HS4x core. "DK" as I understand means "Development Kit". Is it correct name SoC with "DK" abbreviation?

**Comment by abrodkin** [comment]
@kokas-a care to add a reference to the upstream PR which implements it?

**Comment by kokas-a** [comment]
HS4XD support was added to upstream by PR https://github.com/zephyrproject-rtos/zephyr/pull/56683

**Pull request closed without merging**